File: pset6/sentimental/vigenere/vigenere.py
"""
vigenere.py is python program that encrypts messages using Vigenère’s cipher, which improves upon
Caesar's cipher, by encrypting messages using a sequence of keys, from a keyword.
Contrast it with vigenere.c implementation.

CORRECT USAGE:

```
$ python vigenere.py bacon
plaintext: Meet me at the park at eleven am
ciphertext: Negh zf av huf pcfx bt gzrwep oz
```

```
$ python vigenere.py ABC
plaintext:  HELLO
ciphertext: HFNLP
```

INCORRECT USAGE:
```
$ python vigenere.py bacon and eggs
Usage: python vigenere.py k
```

```
$ python vigenere.py bacon and eggs
Usage: python vigenere.py k
```
"""

import sys


def get_ascii_offset(char):
    """
    Function returns ascii offset for upper or lower case character.
    """
    return 65 if char.isupper() else 97


def cipher_char(char, key):
    """
    Function encrypts ASCII upper and lower case char, based on
    integer key provided.
    """
    # Get ASCII offset based on if char is UPPER case or lower case
    ascii_offset = get_ascii_offset(char)
    char_idx = ord(char) - ascii_offset
    trans_char = (char_idx + key) % 26
    char = chr(trans_char + ascii_offset)
    return char


def vigenere(text, keyword, mode):
    """
    Function that implements Vigenère’s algorithm (i.e., cipher) improves upon Caesar's cipher by encrypting
    messages using a sequence of keys (or, put another way, a keyword)

    So,
        - if p is some plaintext (i.e., an unencrypted message)
        - pi is the ith character in p,
        - and kj is a letter in keyword, which is rotated cyclically
            if k is shorter than p, as many times as it takes to encrypt p

    then each letter, ci, in the ciphertext, c, is computed as:
        for kj in keword:
            ci = (pi + kj) % 26
    """
    trans_txt = ""              # stores translated text
    key_index = 0               # counter for keword
    keyword_len = len(keyword)  # keyword length

    # loop through each character in text
    for char in text:
        # a) Check if char is alphabet character or not
        #    if not than continue
        if not char.isalpha():
            trans_txt += char
            continue

        # b) else char is alphabet
        # loop cyclically through keword using
        # key_index
        key_char = keyword[key_index % keyword_len]
        key_index += 1
        # get ASCII offset for UPPER or lower case letter
        # in keyword
        ascii_offset = get_ascii_offset(key_char)
        # convert keyword to integer and subtract
        # ASCII offset
        key = ord(key_char) - ascii_offset
        # check if mode is to decrypt
        if mode == 'd':
            key *= -1  # flips key sign
        trans_txt += cipher_char(char, key)

    # return translated text
    return trans_txt


def main():
    # 1. Parse arguments
    argc = len(sys.argv)  # argument length
    if argc != 2 and argc != 3:
        exit("Usage: python vigenere.py [mode] keyword")

    # 2. Check if the mode is valid to encrypt or decrypt
    mode = sys.argv[1].lower() if argc == 3 else 'e'
    if not mode in 'encrypt e decrypt d'.split():
        exit("Usage: [mode] - either 'encrypt' or 'e' or 'decrypt' or 'd'.")
    if len(mode) > 1:
        mode = mode[0]

    # 3. Check if the keyword is valid sequence of alphabetic letters
    keyword = sys.argv[2] if argc == 3 else sys.argv[1]
    if not keyword.isalpha():
        exit("Usage: keyword - Keyword has to composed entirely of alphabetical characters.")

    # 4. Translate text, depending on mode:

    # 4a) if mode == 'decrypt' or 'd' text is decrypted
    # Prompt user for plain text
    if mode == "d":
        cipher_txt = input("ciphertext: ")
        translated_txt = vigenere(cipher_txt, keyword, mode)
        print("plaintext: ", translated_txt)

    # 4b) else mode == 'encrypt' or 'e', text is encrypted
    else:
        plain_txt = input("plaintext: ")
        translated_txt = vigenere(plain_txt, keyword, mode)
        print("ciphertext: ", translated_txt)


if __name__ == "__main__":
    main()

File: pset2/vigenere/vigenere.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

/**
 * This program implements Vigenere's cipher which improves upon
 * Caesar's cipher, by encrypting messages using a sequence of keys,
 * from a keyword.
 *
 * USAGE:
 * $./vigenere <input string keyword i.e. non-numeric and special characters>
 * plaintext: <input text>
 * ciphertext: <outputs encrypted text>
 *
 **/


// Declare functions here
int get_ascii_offset(char c);
bool is_key_valid(string key);
char encipher_char(char c, int k);


// BEGIN MAIN
int main(int argc, string argv[])
{

    // 1. Get the keyword

    // a) Validate number of CLI arguments
    if (argc != 2)
    {
        printf("Usage: ./vigenere <keyword>\n");
        return 1;

    }

    // b) Check if keyword is valid non-numeric alphabet keyword.
    if (!is_key_valid(argv[1]))
    {
        printf("Keyword has to be non-numeric alphabet keyword.\n");
        return 1;
    }

    // Store keyword in var keyword
    string keyword = argv[1];

    // 2. Get the plain text
    string plain_text = get_string("plaintext: ");


    printf("ciphertext: ");
    int j = 0;
    int keyword_len = strlen(keyword);

    // 3. Encipher
    for (int i = 0, n = strlen(plain_text); i < n; i++)
    {
        // a) Check if char is alphabet character or not
        //    If not than continue
        if (!isalpha(plain_text[i]))
        {
            printf("%c", plain_text[i]);
            continue;
        }

        // b) Encipher the plain_text char
        char key_char = keyword[j++ % keyword_len];
        int ascii_offset = get_ascii_offset(key_char);
        int key = key_char - ascii_offset;
        char c = encipher_char(plain_text[i], key);
        printf("%c", c);
    }

    printf("\n"); // Adds new line
    return 0; // All went well, return success


}
// END MAIN

/**
 * Function encrypts ASCII char, based on the integer key provided.
 * INPUT: char c, int k
 * OUPUT: return encrypted char c
 **/
char encipher_char(char c, int k)
{
    int ascii_offset = get_ascii_offset(c);
    int offset_char = c - ascii_offset;
    offset_char = (offset_char + k) % 26;
    c = (char) offset_char + ascii_offset;
    return c;
}


/**
 * Function cheks whether string key is valid keyword or not.
 * INPUT: string keyword
 * OUTPUT: true or false
 **/
bool is_key_valid(string keyword)
{
    for (int i = 0, n = strlen(keyword); i < n; i++)
    {
        if (!isalpha(keyword[i]))
        {
            return false;
        }
    }

    return true;

}

/**
 * Function returns ascii offset for upper or lower case character
 * INPUT: char c
 * OUTPUT: 65 for UPPER, and 97 for lower
 **/
int get_ascii_offset(char c)
{
    return isupper(c) ? 65 : 97;
}